src/mpid/ch4/include/mpidpre.h

```diff
@@ -282,6 +282,10 @@ typedef struct MPIDI_Devreq_t {
     struct MPIR_Request *anysource_partner_request;
 #endif
 
+    /* Used to track AM request is allocated.
+     * True: ch4.am is being used.
+     * False: ch4.am is not being used. */
+    bool am_in_use;
     union {
```

Comment:
There are quite a few members in this union. Use a single `am_in_use` still takes in assumptions. I think the proper code is to have an enum, e.g. `ch4_req_type` to precisely identify which part of the union is in effect.

Reply:
Good point. Looks like, `am_in_use `is not tracking AM specifically. It is just tracking whether the `req `is there or not. One way to resolve it is change the name e.g: `req_in_use`.  But if are only trying to find whether `req` is allocated or not, then checking `MPIDI_REQUEST(req,req)` in `NULL ` is also fine. Let me look into it a little more and get back.